docs(oss): messages copy pass (#278)
langchain-ai/docs: src/oss/langchain-messages.mdx

message type (e.g. `system`, `user`)
* <Icon icon="folder-closed" size={16} /> [**Content**](#content) - Represents the actual
content of the message (like text, images, audio, documents, etc.)
* <Icon icon="tag" size={16} /> [**Metadata**]() - Optional fields such as response
information, message IDs, and token usage

LangChain provides a standard message type that works across all model
providers, ensuring consistent behavior regardless of what model you're calling.

## Message types
<Columns cols={2}>
    <Card title="System Message" icon="gear" href="#system-message" arrow="true">
        Tells the model how to behave and provide context for interactions
    </Card>
    <Card title="Human Message" icon="user" href="#human-message" arrow="true">
        Represents user input and interactions with the model
    </Card>
    <Card title="AI Message" icon="robot" href="#ai-message" arrow="true">
        Responses generated by the model, including text content, tool calls, and metadata
    </Card>
    <Card title="Tool Message" icon="wrench" href="#tool-message" arrow="true">
        Represents the outputs of tool calls
    </Card>
</Columns>

> ## Using messages

Reviewer: nit: might want to keep it `Basic usage` to match other pages

The simplest way to use messages is to create message objects and pass them to

> a model when calling `.invoke()`.

Reviewer: ```suggestion
a model when [invokating]()`.
```

we could link out to the invocation methods on the models page, since there is more than just `invoke()`

## Overview

Messages are the fundamental unit of context for models in LangChain. They
represent the input and output of models, carrying both content and metadata
that enables a stateful representation of context when interacting with an LLM.

Messages are objects that contain:

* <Icon icon="user" size={16} /> [**Role**](#message-types) - Identifies the
message type (e.g. `system`, `user`)
* <Icon icon="folder-closed" size={16} /> [**Content**](#content) - Represents the actual
content of the message (like text, images, audio, documents, etc.)
* <Icon icon="tag" size={16} /> [**Metadata**]() - Optional fields such as response
information, message IDs, and token usage

LangChain provides a standard message type that works across all model

> providers, ensuring consistent behavior regardless of what model you're calling.

Reviewer: ```suggestion
providers, ensuring consistent behavior regardless of the model being called.
```

messages = [system_msg, human_msg]
response = model.invoke(messages)  # Returns AIMessage
```

### <Icon icon="text" size={20} /> Text prompts

Text prompts are strings - ideal for straightforward generation tasks

> where you don't need to retain conversation history and your request is
> simple.

Reviewer: ```suggestion
where you don't need to retain conversation history.
```

IMO last part is verbose since we already say "straightforward generation tasks"

```python
response = model.invoke("Write a haiku about spring")
```

**Use text prompts when:**
* You have a single, standalone request
* You don't need conversation history
* You want minimal code complexity

### <Icon icon="messages" size={20} /> Message prompts

Alternatively, you can pass in a list of messages to the model by providing

> a list of message objects to the `.invoke()` method.

Reviewer: code example shows `invoke()` so feels duplicate to mention here

```python
from langchain_core.messages import SystemMessage, HumanMessage, AIMessage

messages = [
    SystemMessage("You are a poetry expert"),
    HumanMessage("Write a haiku about spring"),
    AIMessage("Cherry blossoms bloom...")
]
response = model.invoke(messages)
```

**Use message prompts when:**
* Managing multi-turn conversations
* Working with multimodal content (images, audio, files)
* Including system instructions

### <Icon icon="openai" size={20} /> Dictionary format

> You can also pass in a list of dictionaries to the `.invoke()` method, which is

Reviewer: ```suggestion
You can also pass in a list of dictionaries to invocation methods, which is
```

* You want minimal code complexity

### <Icon icon="messages" size={20} /> Message prompts

Alternatively, you can pass in a list of messages to the model by providing
a list of message objects to the `.invoke()` method.

```python
from langchain_core.messages import SystemMessage, HumanMessage, AIMessage

messages = [
    SystemMessage("You are a poetry expert"),
    HumanMessage("Write a haiku about spring"),
    AIMessage("Cherry blossoms bloom...")
]
response = model.invoke(messages)
```

**Use message prompts when:**
* Managing multi-turn conversations
* Working with multimodal content (images, audio, files)
* Including system instructions

### <Icon icon="openai" size={20} /> Dictionary format

You can also pass in a list of dictionaries to the `.invoke()` method, which is
a similar format to what most providers like OpenAI expect.

```python
from langchain_core.messages import SystemMessage, HumanMessage, AIMessage

messages = [
    {"role": "system", "content": "You are a poetry expert"},
    {"role": "user", "content": "Write a haiku about spring"},
    {"role": "assistant", "content": "Cherry blossoms bloom..."}
]
response = model.invoke(messages)
```

## Message types

> ### System Message <Icon icon="gear" size={20} />

Reviewer: IMO icons should go left of the text to follow convention

messages = [system_msg, human_msg]
response = model.invoke(messages)  # Returns AIMessage
```

### <Icon icon="text" size={20} /> Text prompts

Text prompts are strings - ideal for straightforward generation tasks
where you don't need to retain conversation history.

```python
response = model.invoke("Write a haiku about spring")
```

**Use text prompts when:**
* You have a single, standalone request
* You don't need conversation history
* You want minimal code complexity

### <Icon icon="messages" size={20} /> Message prompts

Alternatively, you can pass in a list of messages to the model by providing
a list of message objects when invoking a model.

```python
from langchain_core.messages import SystemMessage, HumanMessage, AIMessage

messages = [
    SystemMessage("You are a poetry expert"),
    HumanMessage("Write a haiku about spring"),
    AIMessage("Cherry blossoms bloom...")
]
response = model.invoke(messages)
```

**Use message prompts when:**
* Managing multi-turn conversations
* Working with multimodal content (images, audio, files)
* Including system instructions

> ### <Icon icon="openai" size={20} /> Dictionary format

Reviewer: TODO fix icon

> You can also pass in a list of dictionaries to invocation methods, which is
> a similar format to what most providers like OpenAI expect.

Reviewer 2: ```suggestion
You can also specify messages directly in OpenAI chat completions format.
```

```python
from langchain_core.messages import SystemMessage, HumanMessage, AIMessage

messages = [
    {"role": "system", "content": "You are a poetry expert"},
    {"role": "user", "content": "Write a haiku about spring"},
    {"role": "assistant", "content": "Cherry blossoms bloom..."}
]
response = model.invoke(messages)
```

## Message types

### <Icon icon="gear" size={20} /> System Message

A `SystemMessage` represent an initial set of instructions that primes the
model's behavior. You can use a system message to set the tone, define the

> model's roles, and establish guidelines for responses.

Reviewer: ```suggestion
model's role, and establish guidelines for responses.
```

<Note>
    The `name` field behavior varies by provider - some use it for user
    identification, others ignore it. To check, refer to the model provider's
    [reference]().
</Note>

### AI Message <Icon icon="robot" size={20} />

An `AIMessage` represents the output of a model invocation. They can include
multimodal data, tool calls, and provider-specific metadata that you can later
access.

<Note>

>     AIMessage objects are returned by the model when calling `.invoke()`, but that

Reviewer: ```suggestion
    `AIMessage`
```

add backticks

>     doesn't mean that's the only place they can be created. Model providers weight

Reviewer: > doesn't mean that's the only place they can be created

good catch

<Note>
    The `name` field behavior varies by provider - some use it for user
    identification, others ignore it. To check, refer to the model provider's
    [reference]().
</Note>

### <Icon icon="robot" size={20} /> AI Message

An `AIMessage` represents the output of a model invocation. They can include
multimodal data, tool calls, and provider-specific metadata that you can later
access.

<Note>
    `AIMessage` objects are returned by the model when invoking it, but that
    doesn't mean that's the only place they can be created. Model providers weight
    different types of messages differently, which means you can use an `AIMessage`

>     to alter model responses, tool calls, and other data.

Reviewer: this feels like two separate points. also in the first sentence, perhaps we have a manual construction example accordion

I also don't like "invoking it"

</Note>

```python
response = model.invoke("Explain AI")
print(type(response))  # <class 'langchain_core.messages.AIMessage'>
```

#### `AIMessage` attributes

<ParamField path="text" type="string">
    The text content of the message.
</ParamField>
<ParamField path="content" type="string | dict[]">

>     The raw content of the message.

Reviewer: if so we may need to modify [this line](https://github.com/langchain-ai/docs/pull/278/commits/5eefdd061d346bac8e2cb67683fd4b5ac6563b65#diff-b3c0a9c4f147108d14890d199d94b7d4fc0560f314fa1fa4b74d6359258bc93cR394)

</ParamField>
<ParamField path="content_blocks" type="types.ContentBlock[]">

>     The standardized content blocks of the message.

Reviewer: let's link to the content blocks header/section

</ParamField>
<ParamField path="tool_calls" type="dict[] | None">
    The tool calls made by the model. (can be empty if no tools are called)
</ParamField>
<ParamField path="id" type="string">

>     A unique identifier for the message (either auto-generated or returned by the provider)

Reviewer: ```suggestion
    A unique identifier for the message (either automatically generated by LangChain or returned in the provider response)
```

</ParamField>
<ParamField path="usage_metadata" type="dict | None">
    The usage metadata of the message, which can contain token counts when available.
</ParamField>
<ParamField path="response_metadata" type="dict | None">

>     The response metadata of the message.

Reviewer: we have a `ResponseMetadata` dict in py

<Note>
    The `name` field behavior varies by provider - some use it for user
    identification, others ignore it. To check, refer to the model provider's
    [reference]().
</Note>

### <Icon icon="robot" size={20} /> AI Message

An `AIMessage` represents the output of a model invocation. They can include
multimodal data, tool calls, and provider-specific metadata that you can later
access.

```python
response = model.invoke("Explain AI")
print(type(response))  # <class 'langchain_core.messages.AIMessage'>
```

<Note>
    `AIMessage` objects are returned by the model when calling it, which contains
    all of the associated metadata in the response. However, that doesn't mean that's
    the only place they can be created/ modified from. Providers weight/contextualize
    types of messages differently, which means it is sometimes advantageous to create
    a new `AIMessage` objects and insert it into the message history as if it came

>     from the model.

Reviewer 2: I'd explicitly say "few shot examples" or similar, here
